### ISSUE TITLE
Better handling or better options for curried function definitions

### ISSUE THREAD

**Issue by mike-robertson** [prettier/prettier#4166]
**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBbAngYQFcAnYgSzgBMAxKTAXkwENiBzARgYD5MAdKXjBasATNz4CYg4QGZxwzgGpmbMctkBuEABoQEAA4wy6ZKBbEIAdwAKLBGmQgmANwhlKOpxmQAzJgBs0OF0AI2ImMABrOBgAZX0IsihWZBhiQmCQSggwXwCg3SSg4hhrcNZsJjzAzIArNAAPACFwqJjYpmw4ABkkuGqCvUIYfWGRAcyE4mLHEKYQ3H9oT31yWAB1dxgAC2QADgAGXVWIIPXw-UdVuGLnft1iOABHQjJHsqYKqqQ-Gt0g7BkVLpTJoJKsfxwACKhAg8AmuiEIU2lB2yBEiPCZH84PwEGwlUcUGg9xAhCCABV5g4fvk4ABfelAA)

Example 1:
**Input:**
```jsx
const myCurriedFn = arg1 => 
  arg2 => 
    arg3 => arg1 + arg2 + arg3;
```

**Output:**
```jsx
const myCurriedFn = arg1 => arg2 => arg3 => arg1 + arg2 + arg3;

```

**Expected behavior:**

I would want this to be preserved because it's much easier to read a curried function like this rather than as one long line. You can easily see how many invocations it requires.

```jsx
const myCurriedFn = arg1 => 
  arg2 => 
    arg3 => arg1 + arg2 + arg3;
```

Example 2:
**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBbAngYQFcAnYgSzgBMAxKTAXkwENiBzARgYD5m2Ambr1YBmQSw6YA1EIHTxwgNwgANCAgAHGGXTJQLYhADuABRYI0yEEwBuEMpRVWMyAGZMANmjiqARsSZgANZwMADK6gFkUKzIMMSE3iCUEGCuHl6qUV7EMMb+rNhMaZ6JAFZoAB4AQv5BIaFM2HAAMlFwxRlqhDDq3XwdiRHE2ZY+TD647tCO6uSwAOr2MAAWyAAcAAyqsxBe8-7qlrNw2dbtqsRwAI6EZJd5TAVFSG4lql7YZLHxiWhRrO44ABFQgQeADVQwcaLSgrZB8SH+Mjuf74CDYQqWKDQc4gQheAAq4wsL3ScAAvuSgA)

**Input:**
```jsx
const myCurriedFn = arg1 => arg2 => arg3 => arg1 + arg2 + arg3;
```

**Output:**
```jsx
const myCurriedFn = arg1 => arg2 => arg3 => arg1 + arg2 + arg3;

```

**Expected behavior:**

I would want this to be forced (space permitting) because it's much easier to read a curried function like this rather than as one long line. You can easily see how many invocations it requires.

```jsx
const myCurriedFn = arg1 => 
  arg2 => 
    arg3 => arg1 + arg2 + arg3;
```

**Comment by qlonik** [comment]
I found this issue while looking for something similar in typescript. It would be nice to have a behavior similar to the original post. Additionally, It would be nice if it was possible to improve this for declared curried functions with explicit type declaration. Also, the problematic printing becomes even more evident when `arrowParens` is set to `always` There are couple problems with the currently printed output, which I will try to cover with different examples.

## Example 1: (reusing code from original post) [Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBbAngYQFcAnYgSzgBMAxKJTACgENiBzARnqkOwCM5iASkwBeAHyMWrAExce-IaInM2AZjl8Bw8Zm6biozFPZKjbaaamrLbEwGozMzA6sBuEABoQEAA4wy6MigLMQQAO4ACiwIaMggTABuEGSUniC8xExgANZwMADKPllkUKzIMMSEcF4AFjDYADYA6jVk8GhFYHD5MW1kCW24cWBosV4laAIwEZms2EzIAGZMDZNeAFZoAB4AQpk5eflM2HAAMiVwSytrIJtb+SWsDXAAioQQ8Fer1SBFxJPEOIwXA+OBoMDkPxpHzkWBNFIwGrIAAcAAYvDCIJMmpkfHEYWCBAlLl4AI7veAzXyxFBMNAAWigcCoVDSxDg5LI7JmTDmCyQy2+Xkm2DI5UqPzQj2ebw+lwF1x+MCYvHhlERyGkXgqTDIDUe+Ag2HmcSg0BJIEIkwAKiqaYKbgkqgBJKCUBAFCFkPwAQTd+WBzy+NzpMGDP0oEDA4a8EEIMB88ek4YAvimgA)
**Input**
```ts
const myCurriedFn: (arg1: number) => (arg2: number) => (arg3: number) => number = arg1 => arg2 => arg3 => arg1 + arg2 + arg3;
```

**Output**
```ts
const myCurriedFn: (
  arg1: number
) => (arg2: number) => (arg3: number) => number = arg1 => arg2 => arg3 =>
  arg1 + arg2 + arg3;
```

**Desired behavior**
I would like for the produced output to be very clear where is the beginning of a type definition and where is the beginning of a function implementation. For example, one possible formatting can be like the following. I'm guessing that this formatting is somewhat unusual and might be hard to get used to, so we can discuss about other approaches.
```ts
const myCurriedFn
  : (arg1: number) => (arg2: number) => (arg3: number) => number
  = arg1 => arg2 => arg3 => arg1 + arg2 + arg3;
```

## Example 2: [Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBbAngYQFcAnYgSzgBMAxKAHSk0yUwAoAbaAcwDkBDbHGpliGAILEuhQbBYAZbsNEwAKrgAOcAJSYAvAD5MDJibacovAXADKcSFEoSpMmPO637lNZp0Hjpkw5ufkEVAAsRR0lpBFdMBQtwyO9tPUN-AJMErgAlOBgSKBT-XUw+NMwAIwqwCvKAairMRrAAbgYGewwcAhJyKloAJn8WIIsQoRFxaJc3CyUMFN90xgCxy0EPaCjnWLmuLYcltIy18w24JOIdmNl47iuvDVSDI1XMpmy8guIi55LMOcFjAAAp8YhWeDECrnQ6UMEQwRQmEPCLXBGQuDQvzvExAqag8GY6GNWF2bYYpFY5qA1GRSn5LGtEAAGhAEHUMDI6GQoHBxAgAHcEQg0MgQHwAG4QMiUVkgSoQsAAa3y1nUfDAZAsyBgxEIcDZYRg2HYAHUIvA0BqwDZRWQuZKHbhxWA0GK2dq0FjQRCuNg+MgAGZ8djetkAKzQAA8AEJK1UwaxWOTauDB0PhkBR6PWbVcdhwACKhAg8AzYcNIA1oix4pgzzQYHInPl6nIsDNspgYWQAA4AAxs9sQb1miHqcXtuDe4iS9NsgCOpfgIIF6jFKD4aAAtFA4FQqPLiHBlyI4Gu+P7A0gQ5W2d7sGRdfqq2h84WS2X07fM1WYHwlRdpQPbIIMbJ6nwZDsPm+AQNgAbilA0ALiAhDeiogGbneWaSgaACSDixNYzZkJyYgONYDaFhWWbbjAtFVpQEBgIxbIQIQMDqJxgyMQAvnxQA)
The produced code becomes even more weird when each argument name and type are pretty long, since prettier then starts to break on opening parenthesis. Also, when `arrowParens` is set to `avoid` (default), printed code actually exceeds the print width. That is because currently prettier does not want to break anywhere after `=>` or after `=`, it only tries to break after each opening parenthesis.

<details><summary>code sample</summary>

**Input**
```ts
const myCurriedFn
  : (longNameFirstArgument: LongFirstType) => 
      (longNameSecondArgument: LongSecondType) =>
        (longNameThirdArgument: LongThirdType) => 
          LongReturnType
  = a => b => c => a + b + c;

const myCurriedFn2
  : (longNameFirstArgument: LongFirstType) => 
      (longNameSecondArgument: LongSecondType) =>
        (longNameThirdArgument: LongThirdType) => 
          LongReturnType
  = longFirstParameter => longSecondParameter => longThirdParameter =>
      longFirstParameter + longSecondParameter + longThirdParameter;
```

**Output**
```ts
const myCurriedFn: (
  longNameFirstArgument: LongFirstType
) => (
  longNameSecondArgument: LongSecondType
) => (longNameThirdArgument: LongThirdType) => LongReturnType = a => b => c =>
  a + b + c;

const myCurriedFn2: (
  longNameFirstArgument: LongFirstType
) => (
  longNameSecondArgument: LongSecondType
) => (
  longNameThirdArgument: LongThirdType
) => LongReturnType = longFirstParameter => longSecondParameter => longThirdParameter =>
  longFirstParameter + longSecondParameter + longThirdParameter;
```
</details>

**Desired behavior**
One possible printing strategy could be as shown in input section. Again, it is very easy to see where is the beginning of type declaration and function implementation. Additionally, it is clear that each curried function accepts only one argument. Here, we do not break on the opening parenthesis as the top priority, but rather after each fat arrow symbol. Another possible printing strategy is to not indent returned functions on each level, but rather to keep them on the same indentation level. Further, if the arguments and types are not so long, then we can fit multiple of those on the same line (as shown in example 1).

<details><summary>code sample</summary>

```ts
// first printing strategy
const myCurriedFn
  : (longNameFirstArgument: LongFirstType) => 
      (longNameSecondArgument: LongSecondType) =>
        (longNameThirdArgument: LongThirdType) => 
          LongReturnType
  = a => b => c => a + b + c;

const myCurriedFn2
  : (longNameFirstArgument: LongFirstType) => 
      (longNameSecondArgument: LongSecondType) =>
        (longNameThirdArgument: LongThirdType) => 
          LongReturnType
  = longFirstParameter => longSecondParameter => longThirdParameter =>
      longFirstParameter + longSecondParameter + longThirdParameter;


// another printing possibility
const myCurriedFn
  : (longNameFirstArgument: LongFirstType) => 
    (longNameSecondArgument: LongSecondType) =>
    (longNameThirdArgument: LongThirdType) => 
    LongReturnType
  = a => b => c => a + b + c;

const myCurriedFn2
  : (longNameFirstArgument: LongFirstType) => 
    (longNameSecondArgument: LongSecondType) =>
    (longNameThirdArgument: LongThirdType) => 
    LongReturnType
  = longFirstParameter => longSecondParameter => longThirdParameter =>
    longFirstParameter + longSecondParameter + longThirdParameter;
```

</details>

## Example 3: [Playground Link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBbAngYQFcAnYgSzgBMAxKAHSk0yUwAoAbaAcwDkBDbHGpliGAILEuhQbBYAZbsNEwAKrgAOcAJSYAvAD5MDJibacovAXADKcSFEoSpMmPO637lNZp0Hjpkw5ufkEVAAsRR0lpBFdMBQtwyO9tPUN-AJMErgAlOBgSKBT-XUw+NMwAIwqwCvKAairMRrAAbgYGewwcAhJyKloAJn8WIIsQoRFxaJc3CyUMFN90xgCxy0EPaCjnWLmuLYcltIy18w24JOIdmNl47iuvDVSDI1XMpmy8guIi55LMOcFjAAAp8YhWeDECrnQ6UMEQwRQmEPCLXBGQuDQvzvExAqag8GY6GNWF2bYYpFY5qA1GRSn5LGtEAAGhAEHUMDI6GQoHBxAgAHcEQg0MgQHx2IK+LgxWzKhCwABrfLWdR8MBkCzIGDEQhwNlhGDYdgAdQi8DQ6rANlFZC5ADd7bhxWA0HKQFq0FjQRCuNg+MgAGaS71sgBWaAAHgAhRUqmDWKxyLVwYOhg0gSNR6xarjsOAARUIEHg6fYYZA6tEWPFMGeaDA5E5rKr5FgprIlBgYWQAA4AAxs9QC72miHqcUjuDe4gOtNsgCOJfgIIF6jFKD4aAAtFA4FQqK3iHBlyI4Gu+P7A0gQxXM97sGQdXqH3mC8XS2nbxm2TA+JUnbdr2SCDH+EJkOweb4BA2ABuKUDQAuICEN6KgAZud6Vg6+oAJIOLE1hNmQnJiA41j1gW5aVtuMDUZmlAQGA9FshAhAwOo7GDPRAC+PFAA)
This example considers example 2 with enabled `arrowParens` to `always`. In this example, we can see that prettier starts breaking on each opening parenthesis. In fact, it chooses which parenthesis to break on given if previous ones fit into the same line or if they didnt (as seen between printed outputs of `myCurriedFn` and `myCurriedFn2`). Regardless of where prettier chooses to break onto the new line, it is very hard to detect that exact `=` from where the implementation of the curried functions starts.

<details><summary>code sample</summary>

**Input**
```ts
const myCurriedFn
  : (longNameFirstArgument: LongFirstType) => 
      (longNameSecondArgument: LongSecondType) =>
        (longNameThirdArgument: LongThirdType) => 
          LongReturnType
  = (a) => (b) => (c) => a + b + c;

const myCurriedFn2
  : (longNameFirstArgument: LongFirstType) => 
      (longNameSecondArgument: LongSecondType) =>
        (longNameThirdArgument: LongThirdType) => 
          LongReturnType
  = (longFirstParameter) => (longSecondParameter) => (longThirdParameter) =>
      longFirstParameter + longSecondParameter + longThirdParameter;
```

**Output**
```ts
const myCurriedFn: (
  longNameFirstArgument: LongFirstType
) => (
  longNameSecondArgument: LongSecondType
) => (longNameThirdArgument: LongThirdType) => LongReturnType = (a) => (b) => (
  c
) => a + b + c;

const myCurriedFn2: (
  longNameFirstArgument: LongFirstType
) => (
  longNameSecondArgument: LongSecondType
) => (longNameThirdArgument: LongThirdType) => LongReturnType = (
  longFirstParameter
) => (longSecondParameter) => (longThirdParameter) =>
  longFirstParameter + longSecondParameter + longThirdParameter;
```

</details>

**Desired behavior**
It would be nice to have it printed in a way similar to example 2 (plus parenthesis around parameters).

## Example 4: [Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBbAngYQFcAnYgSzgBMAxKAHSk0yUwAoAbaAcwDkBDbHGpliGAILEuhQbBYAZbsNEwAKrgAOcAJSYAvAD5MDJibacovAXADKcSFEoSpMmPO637lNZp0Hjpkw5ufkEVAAsRR0lpBFdMBQtwyO9tPUN-AJMErgAlOBgSKBT-XTNFEQwABT5iK3hiNwslDBTfdMYAoIsPaEpq2sF6xq4eh1a0jM7zLiTiPpq6uAb47lmvDW1hvILiIo2JjsymaeaYfsXiTABqTGnR+YH8pevb1Yi588GlgG4GEAAaEAQdQwMjoZCgGrECAAd36CDQyBAfAAbhAyJQASAAEa1MAAa3y1nUfDAZAsyBgxEIcEBYRg2HYAHUIvA0CSwDYEWRQSiebgkWA0IjAeS0EszrUuNg+MgAGZ8djiwEAKzQAA8AEJ4wkwaxWOTkuDyxXKkBq9XWclcdhwACKhAg8BNStpIBJoiWSJgGzQYHIIKx6nIsCZGJgYWQAA4AAyA4MQcVM2rqJHBuDi4go42AgCOjvglWh6kRKD4aAAtFA4FQqFjiHB8yI4EW+NLZUgFa7AeLsGRKdS3WhrbaHU7jZ3TW6YHxsWHKBHkAAmQFUvhkdjW-AQbAypFQaA5kCEcUqWelrtmlE0gCSDli1n9ZBBYgc1h9tpdZvLMC-bsoEBgH+gIQIQMDqGBS5-gAvtBQA)
In this example, consider types are duplicated in the function implementation. While it is not the most plausible way of writing code, where type definition is written twice - in type declaration and in implementation, this example shows that it becomes even harder to distinguish in formatted output where is the beginning of type annotation and where is the beginning of function implementation.

<details><summary>code sample</summary>

**Input**
```ts
const myCurriedFn
  : (longNameFirstArgument: LongFirstType) => 
      (longNameSecondArgument: LongSecondType) =>
        (longNameThirdArgument: LongThirdType) => 
          LongReturnType
  = (longFirstParameter: LongFirstType) => 
      (longSecondParameter: LongSecondType) =>
        (longThirdParameter: LongThirdType): LongReturnType =>
          longFirstParameter + longSecondParameter + longThirdParameter;
```

**Output**
```ts
const myCurriedFn: (
  longNameFirstArgument: LongFirstType
) => (
  longNameSecondArgument: LongSecondType
) => (longNameThirdArgument: LongThirdType) => LongReturnType = (
  longFirstParameter: LongFirstType
) => (longSecondParameter: LongSecondType) => (
  longThirdParameter: LongThirdType
): LongReturnType =>
  longFirstParameter + longSecondParameter + longThirdParameter;
```

</details>

## In conclusion
These examples show that the way prettier is printing curried functions right now is not always the most readable output. There are some questions that still could be considered here. For example, all of the use cases above only consider one parameter to each of the curried functions, but we should consider what to do when multiple parameters are used. 

Additionally, if there is an initial pull request to address this new printing, would the core team consider it?

Thank you for such an awesome project and for keeping working on it! :tada:

**Comment by thorn0** [comment]
Duplicate of #6107
